nixos/modules/system/boot/kernel_config.nix

```diff
@@ -0,0 +1,137 @@
+
+  mergeFalseByDefault = locs: defs:
+    if defs == [] then abort "This case should never happen."
+    else if any (x: x == false) defs then false
+    else true;
+
+  kernelItem = types.submodule {
+    options = {
+      tristate = mkOption {
+        type = types.enum [ "y" "m" "n" null ] // {
+          merge = mergeAnswer [ "y" "m" "n" ];
+        };
+        default = null;
+        internal = true;
+        visible = true;
+        description = ''
+          Use this field for tristate kernel options expecting a "y" or "m" or "n".
+        '';
+      };
+
+      freeform = mkOption {
+        type = types.nullOr types.str // {
+          merge = mergeEqualOption;
+        };
+        default = null;
+        example = ''MMC_BLOCK_MINORS.freeform = "32";'';
+        description = ''
+          Freeform description of a kernel configuration item value.
+        '';
+      };
+
+      optional = mkOption {
+        type = types.bool // { merge = mergeFalseByDefault; };
+        default = false;
+        description = ''
+          Wether option should generate a failure when unused.
+        '';
+      };
+    };
+  };
+
+  mkValue = with lib; val:
+  let
+    isNumber = c: elem c ["0" "1" "2" "3" "4" "5" "6" "7" "8" "9"];
+
+  in
+    if (val == "") then "\"\""
+    else if val == "y" || val == "m" || val == "n" then val
+    else if all isNumber (stringToCharacters val) then val
+    else if substring 0 2 val == "0x" then val
+    else val; # FIXME: fix quoting one day
+
+
+  # generate nix intermediate kernel config file of the form
+  #
+  #       VIRTIO_MMIO m
+  #       VIRTIO_BLK y
+  #       VIRTIO_CONSOLE n
+  #       NET_9P_VIRTIO? y
+  #
+  # Borrowed from copumpkin https://github.com/NixOS/nixpkgs/pull/12158
+  # returns a string, expr should be an attribute set
+  # Use mkValuePreprocess to preprocess option values, aka mark 'modules' as 'yes' or vice-versa
+  # use the identity if you don't want to override the configured values
+  generateNixKConf = exprs:
+  let
+    mkConfigLine = key: item:
+      let
+        val = if item.freeform != null then item.freeform else item.tristate;
+      in
+        if val == null
+          then ""
+          else if (item.optional)
+            then "${key}? ${mkValue val}\n"
+            else "${key} ${mkValue val}\n";
+
+    mkConf = cfg: concatStrings (mapAttrsToList mkConfigLine cfg);
+  in mkConf exprs;
+
+in
+{
+
+  options = {
+
+    intermediateNixConfig = mkOption {
+      readOnly = true;
+      type = types.lines;
+      example = ''
+        USB? y
+        DEBUG n
+      '';
+      description = ''
+        The result of converting the structured kernel configuration in settings
+        to an intermediate string that can be parsed by generate-config.pl to
+        answer the kernel `make defconfig`.
+      '';
+    };
+
+    settings = mkOption {
+      type = types.attrsOf kernelItem;
+      example = literalExample '' with lib.kernel; {
```

Comment:
@teto Note that `with lib.kernel;` is oversimplification. There is no, top-level `kernel` attribute in `lib`.
What is really wanted here, is like
```
  with import ../../../../lib/kernel.nix { inherit (stdenv) lib; inherit version; };
```
Maybe make kernel lib an attribute in `linuxPackages`? My usecase is this (self-contained NixOS config):
```
{ pkgs, lib, ... }: {
  boot.kernelPackages = pkgs.linuxPackages_latest.extend (lib.const (ksuper: {
    kernel =
      with import <nixpkgs/lib/kernel.nix> { inherit lib; version = ksuper.kernel.version; };
      ksuper.kernel.override {
        structuredExtraConfig.SND = no;
      };
  }));
  boot.loader.grub.device = "/nodev";
  fileSystems."/".device = "/nodev";
}
```
And I imagine it like this;
```
{ pkgs, lib, ... }: {
  boot.kernelPackages = pkgs.linuxPackages_latest.extend (lib.const (ksuper: {
    kernel =
      with ksuper.lib; # or even ksuper.kernel.lib
      ksuper.kernel.override {
        structuredExtraConfig.SND = no;
      };
  }));
}
```

Reply:
The description hasn t been updated after lib.kernel got removed. The reason for removal was that i wanted to propose a more holistic version before committing to a public api. So right now structured configs are not practical to use unless you know the implementation (i.e., your solution). I 've started drafting a proposal but i got kinda busy. The idea being to solve https://github.com/NixOS/nixpkgs/issues/41103

Reply:
Ah, maybe make it back?
```
{ lib, version ? null }:

with lib;
rec {
  # Common patterns/legacy
  whenAtLeast = ver: mkIf (versionAtLeast version ver);
  whenOlder   = ver: mkIf (versionOlder version ver);
  # range is (inclusive, exclusive)
  whenBetween = verLow: verHigh: mkIf (versionAtLeast version verLow && versionOlder version verHigh);
 ...
}
```
Then resolve "null" version as "current version" in merger. Even without `whenBetween`, `whenOlder` and `whenAtLeast` this `lib.kernel` is useful (see https://github.com/NixOS/nixpkgs/issues/55415)